addon/v1/types.go

```diff
@@ -0,0 +1,178 @@
+	LatestRelease Release `json:"availableUpdate,omitempty"`
+
+	// CurrentRelease ...
+	CurrentRelease Release `json:"currentVersion,omitempty"`
+
+	// extension contains any additional status information specific to the
+	// operator which owns this status object.
+	// +nullable
+	// +optional
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Extension runtime.RawExtension `json:"extension"`
+}
+
+// ObjectReference contains enough information to let you inspect or modify the referred object.
+type ObjectReference struct {
+	// group of the referent.
+	// +kubebuilder:validation:Required
+	// +required
+	Group string `json:"group"`
+	// resource of the referent.
+	// +kubebuilder:validation:Required
+	// +required
+	Resource string `json:"resource"`
+	// name of the referent.
+	// +kubebuilder:validation:Required
+	// +required
+	Name string `json:"name"`
+}
+
+type ConditionStatus string
+
+// These are valid condition statuses. "ConditionTrue" means a resource is in the condition.
+// "ConditionFalse" means a resource is not in the condition. "ConditionUnknown" means kubernetes
+// can't decide if a resource is in the condition or not. In the future, we could add other
+// intermediate conditions, e.g. ConditionDegraded.
+const (
+	ConditionTrue    ConditionStatus = "True"
+	ConditionFalse   ConditionStatus = "False"
+	ConditionUnknown ConditionStatus = "Unknown"
+)
+
+// AddonStatusCondition represents the state of the operator's
+// managed and monitored components.
+// +k8s:deepcopy-gen=true
+type AddonStatusCondition struct {
+	// type specifies the aspect reported by this condition.
+	// +kubebuilder:validation:Required
+	// +required
+	Type AddonStatusConditionType `json:"type"`
+
+	// status of the condition, one of True, False, Unknown.
+	// +kubebuilder:validation:Required
+	// +required
+	Status ConditionStatus `json:"status"`
+
+	// lastTransitionTime is the time of the last update to the current status property.
+	// +kubebuilder:validation:Required
+	// +required
+	LastTransitionTime metav1.Time `json:"lastTransitionTime"`
+
+	// reason is the CamelCase reason for the condition's current status.
+	// +optional
+	Reason string `json:"reason,omitempty"`
+
+	// message provides additional information about the current condition.
+	// This is only to be consumed by humans.
+	// +optional
+	Message string `json:"message,omitempty"`
+}
+
+// Release represents an OpenShift release image and associated metadata.
+// +k8s:deepcopy-gen=true
+type Release struct {
+	// version is a semantic versioning identifying the update version. When this
+	// field is part of spec, version is optional if image is specified.
+	// +required
+	Version string `json:"version"`
+
+	// imageManifest is list of images will be used for the specific version
+	// this will be use for backward compatability (i.e hub upgraded but agent have not)
+	// +required
+	ImageManifest []Image `json:"imageManifest"`
+}
+
+type Image struct {
+	Name  string `json: "name"`
```

Comment:
The distinction between name and Image needs to be described.  You mean ImagePullSpec perhaps?  Is it required to be a SHA?

Reply:
in the product currently there's a "image-manifest" that contain the information that maps a fixed key to a image. long story...

so the motivation behind this part is that we can record the images used in the current version and when hub upgrade and the hub's image-manifest for the previous version is ripped away we can still manage the agent

Reply:
https://github.com/open-cluster-management/pipeline/blob/2.1-integration/snapshots/manifest-2020-08-24-20-16-08-2.1.0.json

this is one example of such manifest

we use the image-key to lookup the image information that we try to find

Reply:
lets say an agent uses multiple images and image name may change from release to release but the key will be consistent

i was originally imagine that if the agent upgraded if the name changes we can still use the constant key to map this

Reply:
change the fields to ImageKey and ImagePullSpec